pkg/trtl/replica/replica.go

```diff
@@ -0,0 +1,396 @@
+		case replica.Sync_CHECK:
+			// If we're in phase 3 - meaning that the client sent a complete
+			// message, then we ignore check messages to prevent checks bouncing
+			// back and forth from poorly behaved clients.
+			if phase3 {
+				log.Warn().Msg("received check message in phase 3 from client")
+				continue gossip
+			}
+
+			// Mark the object as seen if we're in phase 1 to prevent duplication in phase 2
+			seen.Add(sync.Object)
+
+			// Increment the number of versions seen
+			atomic.AddUint64(&versions, 1)
+
+			// If we're in phase 1 - that means we're receiving version vectors from
+			// the initiating replica, we should compare the incoming version to the
+			// local version on this replica.
+			var local *object.Object
+			if local, err = r.db.Object(sync.Object.Key, options.WithNamespace(sync.Object.Namespace)); err != nil {
+				if err == engine.ErrNotFound {
+					// If this is a not found error, then this object exists on the
+					// initiating replica, but not locally, so request a repair.
+					// Note that we have to set the object version to zero to
+					// indicate that we don't have a version of it, so the client
+					// replica's version is later and it sends a repair message.
+					sync.Object.Version = &object.VersionZero
+					sender.Send(sync)
+				} else {
+					// This is an unhandled error, log and return error information.
+					log.Warn().Err(err).
+						Str("namespace", sync.Object.Namespace).
+						Str("key", b64e(sync.Object.Key)).
+						Msg("failed check sync on initiator: could not fetch object meta")
+
+					// If there is an object error, send a response back to the
+					// initiator's CHECK reqest so that it can be logged there as well.
+					sync.Status = replica.Sync_ERROR
+					sync.Error = err.Error()
+					sender.Send(sync)
+				}
+
+				// CHECK is done here if there was an error.
+				continue gossip
+			}
+
+			// Check which version is later, local or remote
+			switch {
+			case local.Version.IsLater(sync.Object.Version):
+				// Send the local object back to the initiating replica as a repair
+				sync.Status = replica.Sync_REPAIR
+				sync.Object = local
+				sync.Error = ""
+				if ok := sender.Send(sync); ok {
+					atomic.AddUint64(&updates, 1)
+				}
+			case sync.Object.Version.IsLater(local.Version):
+				// Send a check request back to the initiating replica to fetch its version
+				sender.Send(sync)
+			default:
+				// The versions are equal, do nothing
+			}
+
+		case replica.Sync_REPAIR:
+			// Receiving a repaired object, check if it's still later than our local
+			// replica's and if so, save it to disk with an update instead of a put.
+			//
+			// NOTE: we must check if it's later than our local version in case a
+			// concurrent Put (a stomp) or concurrent synchronization updated it.
+			// NOTE: honu.Update performs the version checking in a transaction.
+			if err := r.db.Update(sync.Object, options.WithNamespace(sync.Object.Namespace)); err != nil {
```

Comment:
Looks like this is a place where we'd increment skips or stomps? I guess we'll have to change honu to let us know if there was a PID tiebreaker (stomp) or not (skip)?